include/net/rest_client.h

```diff
@@ -0,0 +1,147 @@
+
+/** @brief Common HTTP status codes */
+enum rest_client_http_status {
+	REST_CLIENT_HTTP_STATUS_OK = 200,
+	REST_CLIENT_HTTP_STATUS_BAD_REQ = 400,
+	REST_CLIENT_HTTP_STATUS_UNAUTH = 401,
+	REST_CLIENT_HTTP_STATUS_FORBIDDEN = 403,
+	REST_CLIENT_HTTP_STATUS_NOT_FOUND = 404,
+};
+
+/**
+ * @brief REST client request and respoonse contexts.
+ *
+ * @details Input parameters for using the REST client API should be filled into
+ *          @ref rest_client_resp_context before calling @ref rest_client_request
+ *          and when the function returns, response-related data can be read from
+ *          @ref rest_client_resp_context structure.
+ */
+struct rest_client_req_context {
+	/** Socket identifier for the connection. When using the default value,
+	 *  the library will open a new socket connection. Default: REST_CLIENT_SCKT_CONNECT.
+	 */
+	int connect_socket;
+
+	/** Defines whether the connection should remain after API call. Default: false. */
+	bool keep_alive;
+
+	/** Security tag. Default: REST_CLIENT_SEC_TAG_NO_SEC. */
+	int sec_tag;
+
+	/** Indicates the preference for peer verification.
+	 *  Initialize to REST_CLIENT_TLS_DEFAULT_PEER_VERIFY
+	 *  and the default (TLS_PEER_VERIFY_REQUIRED) is used.
+	 */
+	int tls_peer_verify;
+
+	/** Used HTTP method. */
+	enum http_method http_method;
+
+	/** Hostname or IP address to be used in the request. */
+	const char *host;
+
+	/** Port number to be used in the request. */
+	uint16_t port;
+
+	/** The URL for this request, for example: /index.html */
+	const char *url;
+
+	/** The HTTP header fields. Similar to the Zephyr HTTP client.
+	 *  This is a NULL-terminated list of header fields. May be NULL.
+	 */
+	const char **header_fields;
+
+	/** Payload/body, may be NULL. */
+	const char *body;
+
+	/** User-defined timeout value for receiving response data.
+	 *  Default: CONFIG_REST_CLIENT_REST_REQUEST_TIMEOUT.
+	 */
+	int32_t timeout_ms;
+
+	/** User-allocated buffer for receiving API response. */
+	char *resp_buff;
+
+	/** User-defined size of resp_buff. */
+	size_t resp_buff_len;
+};
+
+struct rest_client_resp_context {
+	/** Length of HTTP headers + response body/content data. */
+	size_t total_response_len;
+
+	/** Length of response body/content data. */
+	size_t response_len;
+
+	/** Start of response data (the body/content) in resp_buff.*/
+	char *response;
+
+	/** Numeric HTTP status code. */
+	uint16_t http_status_code;
+
+	/** Used socket identifier. Use this for keepalive connections as
+	 *  connect_socket for upcoming requests.
+	 */
+	int used_socket_id;
+
+	/** True if used_socket_id was kept alive and was not closed after the REST request. */
+	int used_socket_is_alive;
+};
+
+/**
+ * @brief REST client request.
+ *
+ * @details This function will block the calling thread until the request completes.
+ *
+ * @param[in] req_ctx Request context containing input parameters to REST request
+ * @param[in,out] resp_ctx Response context for returning the response data.
```

Comment:
@rlubos Should `resp_ctx` be `in,out`, or only `out`? A pointer to the structure is passed into the function, but information inside the structure flows only out from the function. What's the convention here? :)

Reply:
Changed to out parameter